### PR TITLE
Add gpu utils

### PR DESCRIPTION
**Pull request by SergeyTsimfer**
Allows one to use the same line `set_gpu()` in every notebook instead of manually changing the number of used device

### CODE REVIEW

File: batchflow/utils.py [hunk 1]
```diff
@@ -406,3 +406,61 @@ def read_data_from(src, **kwargs):
         data = data[0]
 
     return data
+
+def get_available_gpus(n=1, fraction_memory=0.9, max_processes=2):
```
Comment (cdtn) [inline batchflow/utils.py:410]:
```suggestion
def get_available_gpus(n=1, min_free_memory=0.9, max_processes=2):
```

File: batchflow/utils.py [hunk 2]
```diff
@@ -406,3 +406,61 @@ def read_data_from(src, **kwargs):
...
+        num_processes = len(nvidia_smi.nvmlDeviceGetComputeRunningProcesses(handle))
+
+        if (fraction_free > fraction_memory) & (num_processes <= max_processes):
+            available_devices.append(i)
+
+    if len(available_devices) < n:
+        raise ValueError(f'Not enough free devices: requested {n}, found {len(available_devices)}')
+    return available_devices[:n]
+
+def set_gpus(n=1, fraction_memory=0.9, max_processes=2):
```
Comment (cdtn) [inline batchflow/utils.py:448]:
```suggestion
def use_available_gpus(n=1, fraction_memory=0.9, max_processes=2):
```

**Review by cdtn — CHANGES_REQUESTED** [review]
Add option to choose the hottest gpu 🔥

File: batchflow/utils.py [hunk 3]
```diff
@@ -406,3 +406,61 @@ def read_data_from(src, **kwargs):
...
+    fraction_memory : float
+        Minimum percentage of free memory on a device to consider it free.
+    max_processes : int
+        Maximum amount of computed processes on a device to consider it free.
+
+    Returns
+    -------
+    List with indices of availble GPUs
+    """
+    if nvidia_smi is None:
```
Comment (anton-br) [inline batchflow/utils.py:433]: forgot to import `nvidia_smi` maybe?

File: batchflow/utils.py [hunk 4]
```diff
@@ -406,3 +406,61 @@ def read_data_from(src, **kwargs):
...
+    max_processes : int
+        Maximum amount of computed processes on a device to consider it free.
+    """
+    if 'CUDA_VISIBLE_DEVICES' in os.environ.keys():
+        raise ValueError('`CUDA_VISIBLE_DEVICES` is already set!')
+
+    devices = get_available_gpus(n=n, fraction_memory=fraction_memory, max_processes=max_processes)
+    str_devices = ','.join(str(i) for i in devices)
+    os.environ['CUDA_VISIBLE_DEVICES'] = str_devices
+    return str_devices
```
Comment (anton-br) [inline batchflow/utils.py:473]: suggest to return list of devices

File: batchflow/utils.py [hunk 5]
```diff
@@ -436,23 +436,29 @@ def get_available_gpus(n=1, min_free_memory=0.9, max_processes=2):
...
+
+        if log:
+            print(f'Device {i} | Free memory: {fraction_free:4.2f} | '
+                  f'Number of running processes: {num_processes:>2} | Free: {consider_available}')
 
     if len(available_devices) < n:
         raise ValueError(f'Not enough free devices: requested {n}, found {len(available_devices)}')
-    return available_devices[:n]
+    available_devices = np.array(available_devices)[np.argsort(memory_usage)[::-1]]
+    return sorted(available_devices[:n])
```
Comment (a-arefina) [inline batchflow/utils.py:459]: why sort again?

File: batchflow/utils.py [hunk 6]
```diff
@@ -436,23 +436,29 @@ def get_available_gpus(n=1, min_free_memory=0.9, max_processes=2):
...
 
-def set_gpus(n=1, min_free_memory=0.9, max_processes=2):
+def set_gpus(n=1, min_free_memory=0.9, max_processes=2, log=False):
```
Comment (a-arefina) [inline batchflow/utils.py:461]: since log=False still doesn't silent this function absolutely maybe it's better to use `verbosity` ?

**Review by a-arefina — APPROVED** [review]

File: batchflow/utils.py [hunk 7]
```diff
@@ -406,3 +412,71 @@ def read_data_from(src, **kwargs):
...
+    """
+    if 'CUDA_VISIBLE_DEVICES' in os.environ.keys():
+        raise ValueError('`CUDA_VISIBLE_DEVICES` is already set!')
+
+    devices = get_available_gpus(n=n, min_free_memory=min_free_memory, max_processes=max_processes, log=log)
+    str_devices = ','.join(str(i) for i in devices)
+    os.environ['CUDA_VISIBLE_DEVICES'] = str_devices
+
+    newline = "\n" if log else ""
+    print(f'{newline}`CUDA_VISIBLE_DEVICES` set to "{str_devices}"')
```
Comment (roman-kh) [inline batchflow/utils.py:481]: Remove `print` or make it optional

File: batchflow/utils.py [hunk 8]
```diff
@@ -406,3 +412,71 @@ def read_data_from(src, **kwargs):
...
+        fraction_free = info.free / info.total
+        num_processes = len(nvidia_smi.nvmlDeviceGetComputeRunningProcesses(handle))
+
+        consider_available = (fraction_free > min_free_memory) & (num_processes <= max_processes)
+        if consider_available:
+            available_devices.append(i)
+            memory_usage.append(fraction_free)
+
+        if log:
+            print(f'Device {i} | Free memory: {fraction_free:4.2f} | '
```
Comment (roman-kh) [inline batchflow/utils.py:453]: It's confusing when `log` prints

File: batchflow/__init__.py [hunk 9]
```diff
@@ -18,7 +18,7 @@
 from .decorators import action, inbatch_parallel, parallel, any_action_failed, mjit, deprecated, apply_parallel
 from .exceptions import SkipBatchException, EmptyBatchSequence, StopPipeline
 from .sampler import Sampler, ConstantSampler, NumpySampler, HistoSampler, ScipySampler
-from .utils import save_data_to, read_data_from
+from .utils import save_data_to, read_data_from, get_available_gpus, set_gpus
```
Comment (roman-kh) [inline batchflow/__init__.py:21]: Perhaps it would be more convenient to move it to `utils_gpu`

File: batchflow/utils.py (hunk 5 above)
Comment (SergeyTsimfer) [reply batchflow/utils.py:459]: The first sort returns the list of GPUs, sorted by free memory, so we take 'the most free devices'
The second sort makes this list an increasing sequence of device indices

File: batchflow/utils.py (hunk 7 above)
Comment (SergeyTsimfer) [reply batchflow/utils.py:485]: The behavior of `%env CUDA_VISIBLE_DEVICES=#` is to print on success, and I wanted to replicate this. Making it optional would introduce an additional parameter (as you can't merge this with `verbose`) that would probably never be used anyways

**Comment by roman-kh** [comment]
Resolve the conflict and move it to utils_notebook

**Pull request merged**